packages/state-transition/src/cache/epochCache.ts

```diff
@@ -309,8 +309,10 @@ export class EpochCache {
       if (cachedPreviousShuffling == null && isActiveValidator(validator, previousEpoch)) {
         previousActiveIndices.push(i);
       }
-      if (cachedCurrentShuffling == null && isActiveValidator(validator, currentEpoch)) {
-        currentActiveIndices.push(i);
+      if (isActiveValidator(validator, currentEpoch)) {
+        if (cachedCurrentShuffling == null) {
+          currentActiveIndices.push(i);
+        }
         // We track totalActiveBalanceIncrements as ETH to fit total network balance in a JS number (53 bits)
         totalActiveBalanceIncrements += effectiveBalanceIncrements[i];
```

Comment:
One idea to prevent those kind of issues slipping through gradually could be to physically separate blocks that are not logically related.

e.g.

```js
//1. First everything `activeIndices` related

// we only need to track active indices for previous, current and next epoch if we have to compute shufflings
// skip doing that if we already have cached shufflings
const validatorCurrentlyActive = isActiveValidator(validator, currentEpoch);
if (cachedPreviousShuffling == null && isActiveValidator(validator, previousEpoch)) {
  previousActiveIndices.push(i);
}
if (validatorCurrentlyActive) {
  if (cachedCurrentShuffling == null) {
    currentActiveIndices.push(i);
  }
}
if (cachedNextShuffling == null && isActiveValidator(validator, nextEpoch)) {
  nextActiveIndices.push(i);
}

//2. Then extra specific logic
if (validatorCurrentlyActive) {
  // We track totalActiveBalanceIncrements as ETH to fit total network balance in a JS number (53 bits)
  totalActiveBalanceIncrements += effectiveBalanceIncrements[i];
}
```